src/app/api/game-sessions/route.ts

```diff
@@ -20,21 +92,297 @@ export async function GET() {
+  } catch (error) {
+    if (error instanceof z.ZodError) {
+      return NextResponse.json(error.issues, { status: 400 });
+    }
+    return new Response("Internal Server Error", { status: 500 });
+  }
+}
+
+export async function PUT(req: NextRequest) {
+  try {
+    const user = await getCurrentUser();
+    if (!user) {
+      return new Response("ERROR: Unauthorized request", { status: 401 });
+    }
+    if (user.role != "admin") {
+      return new Response("ERROR: No valid permissions", { status: 403 });
+    }
+
+    const date = req.nextUrl.searchParams.get("date");
+    const gameSessionDate = z.string().date().parse(date);
+
+    const gameSessionToUpdate = updateGameSessionSchema.parse(await req.json());
+
+    // Check the game session to update is valid
+    if (gameSessionToUpdate.startTime > gameSessionToUpdate.endTime) {
+      return new Response("Start time must be less than end time", {
+        status: 400,
+      });
+    }
+
+    if (gameSessionToUpdate.casualCapacity > gameSessionToUpdate.capacity) {
+      return new Response("Casual capacity must be less than capacity", {
+        status: 400,
+      });
+    }
+
+    // Find the semester that the date is in
+    const semester = await db.query.semesters.findFirst({
+      where: or(
+        and(
+          lte(semesters.startDate, gameSessionDate),
+          gt(semesters.breakStart, gameSessionDate),
+        ),
+        and(
+          lt(semesters.breakEnd, gameSessionDate),
+          gte(semesters.endDate, gameSessionDate),
+        ),
+      ),
+    });
+
+    if (!semester) {
+      return new Response("No semester found for this date", { status: 400 });
+    }
+
+    // Calculate the booking open and close times and convert to NZ Standard Time
+    let bookingOpen = parse(
+      `${semester.bookingOpenDay} ${semester.bookingOpenTime}`,
+      "iiii HH:mm:ss",
+      new Date(gameSessionDate),
+    );
+
+    // If the booking open time is after the game session date, set it back a week
+    if (bookingOpen > new Date(gameSessionDate)) {
+      bookingOpen.setDate(bookingOpen.getDate() - 7);
+    }
+
+    bookingOpen = fromZonedTime(bookingOpen, "Pacific/Auckland");
+
+    let bookingClose = parse(
+      `${gameSessionDate} ${gameSessionToUpdate.startTime}`,
+      "yyyy-MM-dd HH:mm:ss",
+      new Date(gameSessionDate),
+    );
+
+    bookingClose = fromZonedTime(bookingClose, "Pacific/Auckland");
+
+    const gameSessionToInsert = insertGameSessionSchema.parse({
+      ...gameSessionToUpdate,
+      date: gameSessionDate,
+      bookingOpen: bookingOpen,
+      bookingClose: bookingClose,
+    });
+
+    // If the game session exists, update it and return
+    const [updatedGameSession] = await db
+      .update(gameSessions)
+      .set(gameSessionToUpdate)
+      .where(eq(gameSessions.date, gameSessionDate))
+      .returning();
+
+    if (updatedGameSession) {
+      return new Response(null, { status: 204 });
+    }
+
```

Comment:
I would put this on line 311 (before the bookingOpen calculations) to better group the different responsibilities